library.py

```diff
@@ -0,0 +1,174 @@
+import torch
+import inspect
+import functools
+
+"""
+These are changes that we'll add to PyTorch
+"""
+
+
+class OpDef:
+    @classmethod
+    def build(cls):
+        return register(cls)
+
+
+OPERATOR_REGS = {}
+
+
+def register(opdef):
+    name = opdef.__name__
+
+    frm = inspect.stack()[2]
+    mod = inspect.getmodule(frm[0])
+    ns = mangle_module(mod.__name__)
+
+    check_allowed_attrs(opdef)
+    qualname = f"{ns}::{name}"
+    lib = get_library_allowing_overwrite(ns, name)
+    lib.define(f"{name}{opdef.schema}")
+    op = torch._library.utils.lookup_op(qualname)
+
+    impl_method = getattr(opdef, "impl", None)
+    if impl_method is not None:
+        properties = getattr(impl_method, "__properties__", {})
+        for device in properties["device_types"]:
+            register_backend(lib, name, impl_method, device)
+
+    impl_cpu_method = getattr(opdef, "impl_cpu", None)
+    if impl_cpu_method is not None:
+        register_backend(lib, name, impl_cpu_method, "CPU")
+
+    impl_cuda_method = getattr(opdef, "impl_cuda", None)
+    if impl_cuda_method is not None:
+        register_backend(lib, name, impl_cuda_method, "CUDA")
+
```

Comment:
I guess in the final version we'll want to register error kernels for all other backends and features)?